test/boost/commitlog_test.cc

```diff
@@ -1089,3 +1089,54 @@ BOOST_AUTO_TEST_CASE(test_commitlog_segment_descriptor) {
     }
 }
 
+SEASTAR_TEST_CASE(test_commitlog_max_data_lifetime) {
+    commitlog::config cfg;
+
+    constexpr auto max_size_mb = 1;
+
+    cfg.commitlog_segment_size_in_mb = max_size_mb;
+    cfg.commitlog_total_space_in_mb = 2 * max_size_mb * smp::count;
+    cfg.commitlog_sync_period_in_ms = 10;
+    cfg.commitlog_data_max_lifetime_in_seconds = 2;
+    cfg.allow_going_over_size_limit = false;
+    cfg.use_o_dsync = true; // make sure we pre-allocate.
+
+    tmpdir tmp;
+    cfg.commit_log_location = tmp.path().string();
+    auto log = co_await commitlog::create_commitlog(cfg);
+
+    rp_set rps;
+    // uncomment for verbosity
+    // logging::logger_registry().set_logger_level("commitlog", logging::log_level::debug);
+
+    auto uuid = make_table_id();
+    auto size = log.max_record_size();
+
+    std::unordered_set<cf_id_type> ids;
+
+    condition_variable cond;
+
+    auto r = log.add_flush_handler([&](cf_id_type id, replay_position pos) {
+        log.discard_completed_segments(id, rps);
+        ids.insert(id);
+        cond.signal();
+    });
+
+    rp_handle h = co_await log.add_mutation(uuid, size, db::commitlog::force_sync::no, [&](db::commitlog::output& dst) {
+        dst.fill('1', size);
+    });
+    h.release();
+
+    // should not be signaled yet. 
+    BOOST_REQUIRE(!ids.contains(uuid));
+
+    while (!ids.contains(uuid)) {
+        co_await cond.wait(30s);
```

Comment:
why so long in a unit test?
can we shorten the intervals and wait 1s at a time.

also, if the test fails we should hang here forever, so fail the test if we waited for unreasonable time (and maybe it should depend on the build mode).

if we can use the manual clock and make this test deterministic and independent of the system clock it would even be better.

Reply:
Because we had so much issues with machines almost pausing in debug mode etc previously. Paranoia. 
But you are right, we should only wait max 2 times or so.